cpp/ql/src/Likely Bugs/Memory Management/UsingExpiredStackAddress.ql

```diff
@@ -0,0 +1,247 @@
+/**
+ * @name Use of expired stack-address
+ * @description Accessing the stack-allocated memory of a function
+ *              after it has returned can lead to memory corruption.
+ * @kind problem
+ * @problem.severity error
+ * @security-severity 9.3
+ * @precision high
+ * @id cpp/using-expired-stack-address
+ * @tags reliability
+ *       security
+ *       external/cwe/cwe-825
+ */
+
+import cpp
+// We don't actually use the global value numbering library in this query, but without it we end up
+// recomputing the IR.
+import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+import semmle.code.cpp.ir.IR
+
+predicate instructionHasVariable(VariableAddressInstruction vai, StackVariable var, Function f) {
+  var = vai.getASTVariable() and
+  f = vai.getEnclosingFunction() and
+  // Pointer-to-member types aren't properly handled in the dbscheme.
+  not vai.getResultType() instanceof PointerToMemberType and
+  // Rule out FPs caused by extraction errors.
+  not any(ErrorExpr e).getEnclosingFunction() = f
+}
+
+/**
+ * Holds if `source` is the base address of an address computation whose
+ * result is stored in `address`.
+ */
+predicate stackPointerFlowsToUse(Instruction address, VariableAddressInstruction source) {
```

Comment:
Could this just be `DataFlow::localFlow` or `TaintTracking::localTaint`?

Reply:
It could, but that would also give us flow through phi nodes. I've seen multiple FPs from results like:
```cpp
int* global1;
void f(bool b) {
  int local;
  int *p1, *p2;
  if(b) {
    p1 = &local;
  } else {
    p1 = p2;
  }
  // ...
  if(!b) {
    global = p1;
  }
}
```
where the address of the local doesn't actually escape due to complex path conditions.